qcodes/instrument/abstract.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Type
+from qcodes import Parameter
+from qcodes.instrument.base import InstrumentBase
+from qcodes.instrument.parameter import _BaseParameter
+
+
+class AbstractParameter(Parameter):
+    """
+    This is a trivial subclass of 'Parameter' to signal
+    that this parameters *must* be overridden in
+    subclasses of abstract instruments.
+    """
+
+
+class AbstractParameterException(Exception):
+    """
+    Errors with abstract parameter utilization should
+    raise this exception class.
+    """
+
+
+def abstract_instrument(klass: Type[InstrumentBase]) -> Type[InstrumentBase]:
+    """
+    A class decorator to create an abstract instrument. Abstract
+    instruments are allowed to have abstract parameters, but
+    their subclasses do not. This works by replacing the
+    'add_parameter' and '__init_subclass__' methods of the class
+    dynamically.
+
```

Comment:
I do not see the benefit of the existence of abstract instrument. I would simply put all the logic inside the InstrumentBase class  such that all instruments can have abstract parameters. The requirement to use ABC or ABC metaclass for abstract methods is imho only due to historical reasons  so lets not repeat that